module/VuFind/src/VuFind/ILS/Connection.php

```diff
@@ -1227,4 +1288,82 @@ public function __call($methodName, $params)
             'Cannot call method: ' . $this->getDriverClass() . '::' . $methodName
         );
     }
+
+    /**
+     * Get data for an ILS method from shared or session cache
+     *
+     * @param string $methodName The name of the called method.
+     * @param array  $params     Array of passed parameters.
+     *
+     * @return ?array
+     */
+    protected function getCachedData($methodName, $params)
+    {
+        $cacheLifeTime = $this->cacheLifeTime[$methodName] ?? null;
+        $cacheStorage = $this->cacheStorage[$methodName] ?? null;
+        if (!$cacheLifeTime || !$cacheStorage) {
+            return null;
+        }
+        $cacheKey = $methodName . md5(serialize($params));
+        if ('shared' === $cacheStorage) {
+            return $this->getSharedCachedData($cacheKey);
+        }
+        if ($this->sessionCache && ($entry = $this->sessionCache[$cacheKey] ?? null)) {
+            if (time() - $entry['ts'] <= $cacheLifeTime) {
+                return $entry['data'];
+            }
+            unset($this->sessionCache[$cacheKey]);
+        }
+        return null;
+    }
+
+    /**
+     * Put data for an ILS method to shared or session cache.
+     *
+     * @param string $methodName The name of the called method.
+     * @param array  $params     Array of passed parameters.
+     * @param mixed  $data       Data to cache
+     *
+     * @return void
+     */
+    protected function putCachedData($methodName, $params, $data): void
+    {
+        $cacheLifeTime = $this->cacheLifeTime[$methodName] ?? null;
+        $cacheStorage = $this->cacheStorage[$methodName] ?? null;
+        if (!$cacheLifeTime || !$cacheStorage) {
+            return;
+        }
+        $cacheKey = $methodName . md5(serialize($params));
+        if ('shared' === $cacheStorage) {
+            $this->putSharedCachedData($cacheKey, $data);
+            return;
+        }
+        if ($this->sessionCache) {
+            $this->sessionCache[$cacheKey] = [
+                'ts' => time(),
+                'data' => $data,
+            ];
+        }
+    }
+
+    /**
+     * Default method -- pass along calls to the driver if available; return
+     * false otherwise. This allows custom functions to be implemented in
+     * the driver without constant modification to the connection class.
+     *
+     * @param string $methodName The name of the called method.
+     * @param array  $params     Array of passed parameters.
+     *
+     * @throws ILSException
+     * @return mixed             Varies by method (false if undefined method)
+     */
+    public function __call($methodName, $params)
+    {
+        if ($entry = $this->getCachedData($methodName, $params)) {
+            return $entry['data'];
```

Comment:
In line 1313, you return `$entry['data']` and now here you're also referencing `$entry['data']`. I think this is correct and there really is an `$entry['data']['data']` in play... but that's a bit confusing. Might there be a way to make this a little easier to follow?

Reply:
Yes, renamed keys and arrays, and added comments. Maybe better now?